rclcpp/include/rclcpp/subscription_options.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__SUBSCRIPTION_OPTIONS_HPP_
+#define RCLCPP__SUBSCRIPTION_OPTIONS_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+
+struct SubscriptionEventCallbacks
+{
+  // NOTE(emersonknapp): empty placeholder for future additions of QoS Event callbacks
+};
+
+template<typename Alloc = std::allocator<void>>
+class SubscriptionOptions
```

Comment:
**High-level comment about this class and the other ones**

If possible, it'd be great to see if we can avoid setters in classes as it increases the complexity of the class for little benefits.
We could for instance write:

```c++
auto opts = SubscriptionOptionsBuilder().set_foo(42).set_bar("aaa").build()
// opts is completely const / immutable here.
```

Copy and assignment constructors are also best avoided at all if possible.
For those non performance-critical classes, we can avoid boilerplate by just relying on unique_ptr + move.

Copy ctor + mutability is deadly when combined with multi-threading and accidental copies (you thought you modified the original object? well actually that's a temporary)

This applies also to the other classes as far as I could see.